### PR TITLE
[MM-12304] Add channel search support in mattermost cli

### PR DESCRIPTION
**Pull request by adzimzf**
#### Summary
add command search on channel group.
it could search channel by name or id in general
and search channel with team by team name or id

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/9491
https://mattermost.atlassian.net/browse/MM-12304

#### Checklist
- [x] Added or updated unit tests (required for all new features)

### CODE REVIEW

**Comment by hanzei** [comment]
Thank you very much @adzimzf for the PR! Is this ready for review?

**Comment by adzimzf** [comment]
hi @hanzei,
sure it's ready.
I use loop to get channel in all team, since there's no method getChannel that has only channel name param. getChannelBy name has 2 param that's channel name and team id.

**Comment by adzimzf** [comment]
Hi @hanzei,

Since this is my first PR in open source project I am very excited on this PR to approve.

Is there any mistakes on my changes? or any comments?

**Comment by hanzei** [comment]
Hey @adzimzf,

Sorry for taking a bit longer as usual to review your PR. Your surely did nothing wrong. Let me explain the process of the review:
@wiersgallak will review your PR from a UX point of view. After she approves two devs will review. And then we can merge your PR. I know this is a long process, but it ensure the  quality of our code and software. Looking forward to merging this!

> I use loop to get channel in all team, since there's no method getChannel that has only channel name param. getChannelBy name has 2 param that's channel name and team id.

That's correct. If no team is specified, we want to search through all teams.

> is that right searching feature only result channel name?? or should show another information.

@wiersgallak Can you take a look at this questions?

**Comment by adzimzf** [comment]
Hi @hanzei,
No problem.
Sorry, I am new comer in open source project, so I don't know the flow.
Thanks for your explanation. I totally agree that code quality is very important.
Feel free to let me know if any mistakes in my changes.

**Comment by wiersgallak** [comment]
Hello @adzimzf and thank you for your work on this.  I would recommend that we follow the information we are returning in the search teams command and provide: 
channel name, channel display name, channel id and "archived" when the channel has been archived. 

Please let me know if I can assist with any other questions.

**Comment by adzimzf** [comment]
Hi @wiersgallak,
Thanks for info, I'll update the PR as soon as possible

**Comment by adzimzf** [comment]
Hi @hanzei,
how about this ?

**Comment by hanzei** [comment]
Hey @adzimzf,

I highly appreciate the time and afford you put into this and I understand that you want this to be merged as soon as possible.

Mattermost hosted a conference over the last week and most of the employees didn't had time to look into PR. Normal work will begin today and I'm sure your PR will get reviewed soon.


@wiersgallak Can you take a look at this one?

**Comment by adzimzf** [comment]
Hi @hanzei,
I am very sorry, it seem I was force in this PR.

**Comment by hanzei** [comment]
@adzimzf There is nothing to be sorry about. It's great if you keep the process running!

**Comment by hanzei** [comment]
@mgdelacroix @jespino Can you please review, when you have the time?

File: cmd/mattermost/commands/channel.go [hunk 1]
```diff
@@ -533,3 +546,62 @@ func renameChannelCmdF(command *cobra.Command, args []string) error {
 
 	return nil
 }
+
+func searchChannelCmdF(command *cobra.Command, args []string) error {
+
+	var (
```
Comment (jespino) [inline cmd/mattermost/commands/channel.go]: We don't define the vars at the top of the function, we normally define them, just before the first usage.

File: cmd/mattermost/commands/channel.go [hunk 2]
```diff
@@ -533,3 +546,62 @@ func renameChannelCmdF(command *cobra.Command, args []string) error {
...
+		teamId    string
+	)
+
+	a, err := InitDBCommandContextCobra(command)
+	if err != nil {
+		return errors.Wrap(err, "failed to InitDBCommandContextCobra")
+	}
+	defer a.Shutdown()
+
+	teams, aErr := a.GetAllTeams()
```
Comment (jespino) [inline cmd/mattermost/commands/channel.go]: I don't want to get all the teams, all the time, I prefer to get the team by name if you pass the team name, and get all the teams if not.

File: cmd/mattermost/commands/channel_test.go [hunk 3]
```diff
@@ -131,3 +131,16 @@ func TestRenameChannel(t *testing.T) {
...
 }
+
+func Test_searchChannelCmdF(t *testing.T) {
+	th := Setup().InitBasic()
+	defer th.TearDown()
+
+	channel := th.CreatePublicChannel()
+
+	th.CheckCommand(t, "channel", "search", channel.Name)
+	th.CheckCommand(t, "channel", "search", channel.Name+"404")
```
Comment (jespino) [inline cmd/mattermost/commands/channel_test.go]: You are checking that the command runs, and doesn't fail, but you are not checking that the command does what is expected. Here is needed to test the output too.

**Comment by adzimzf** [comment]
Thanks for your feedback @jespino
I'll update asap

**Comment by adzimzf** [comment]
Hi @jespino,
I have some update.

File: cmd/mattermost/commands/channel.go [hunk 4]
```diff
@@ -134,6 +145,7 @@ func init() {
 	ModifyChannelCmd.Flags().String("username", "", "Required. Username who changes the channel privacy.")
 
 	ChannelRenameCmd.Flags().String("display_name", "", "Channel Display Name")
+	SearchChannelCmd.Flags().Bool("team", false, "Team name or ID")
```
Comment (jespino) [inline cmd/mattermost/commands/channel.go]: This mustn't be a Bool, it should be a "String" with the team name or id. this way you can use any order, like `channel search myChannel --team myTeam`.

**Comment by jespino** [comment]
@adzimzf sorry the delay. I think we are almost there, the rest of the code looks good to me. but you are using a `Bool` flag to pass the team, makes more sense to use a `String` flag, and use the flag value, and then, the channel name will be always args[0].

**Comment by adzimzf** [comment]
Hi @jespino,
I have some update.

**Comment by jespino** [comment]
@adzimzf really sorry the delay (again), now looks good to me.

**Comment by hanzei** [comment]
Thanks you very much for the contribution @adzimzf! :tada:

**Pull request merged**